src/java.base/share/classes/java/lang/String.java

```diff
@@ -695,14 +701,14 @@ private String(Charset charset, byte[] bytes, int offset, int length) {
                 throw new Error(x);
             }
             if (COMPACT_STRINGS) {
```

Comment:
Now we can call other constructors of String, so instead of having this manual handling of char[], we can call `return new String(ca, 0, caLen, null)`. Same for the `clen` version above.

Reply:
```java
    private String(char[] value, int off, int len, Void sig) {
        if (len == 0) {
            this.value = "".value;
            this.coder = "".coder;
            return;
        }
        if (COMPACT_STRINGS) {
            byte[] val = StringUTF16.compress(value, off, len);
            this.coder = StringUTF16.coderFromArrayLen(val, len);
            this.value = val;
            return;
        }
        this.coder = UTF16;
        this.value = StringUTF16.toBytes(value, off, len);
    }
```

This constructor has additional processing logic for len == 0. 

ValueObject may be used here in the future. I think it is better to keep the original structure.

Reply:
I thought this would both increase cleaniness and reduce code size of the decode method. @minborg What do you think? This also removes one of the switches on COMPACT_STRING.

Reply:
That zero-length check in `String(char[], int, int, Void)` was done as an optimization, too, so would be good if we leveraged it in paths that currently miss out for consistency. It was done maybe not so much for speed but to reduce memory and allocation pressure (empty `String`s are surprisingly common in many apps).